torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -644,6 +644,12 @@ bool ProcessGroupNCCL::WorkNCCL::wait(std::chrono::milliseconds timeout) {
       static_cast<int>(devices_.size())); // worldSize
   synchronizeInternal(timeout);
   // Always return true, because abort API is not implemented.
+  if (debug_level() == DebugLevel::Detail) {
+    auto hashValue = hashTensors(*outputs_);
```

Comment:
i'm confused about the terminology.  In work.wait() we are just sure that we have launched the kernel to the cuda stream, so we do not know about the tensor outputs yet right?  why not call these 'inputs'?

also, why do we have logging in both 'collective' and 'wait' - would these give different results?  (or maybe that part is intended, just to prove we didn't corrupt between creating and launching?)

In any case, the print statement coming from `PRINT_COLLECTIVE_HASH_SIGNATURE` may be confusing to users if the exact same print shows up twice for one collective.  it might be worth adding an argument for "phase" so we can print "creating" vs "launching" collective or something

Reply:
Sure, I can add stages here. But I remember @fegin mentioned that he saw discrepancies between here and inside NCCLPG.

Reply:
@wconstab The hash function calls `cudaMemcpy` so it will actually trigger a synchronization. `wait()` will actually wait for the collective to finish if the hash function is called. That's why it is very slow.

Reply:
oh i get it now.

@fduwjj I was confused earlier, but now it makes sense why there were 2 prints, and the one in 'wait' should say "after" and the one in collective should say "before".  does this make sense?  I think the latest code only has one print, but let me review again to double check.